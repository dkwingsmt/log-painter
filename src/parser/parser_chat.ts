import compact from 'lodash/compact';
import trim from 'lodash/trim';
import flow from 'lodash/flow';

export interface ParsedPlayer {
  name: string;
  number?: string;
  title?: string;
}

export interface ParsedLine {
  time?: string;
  player: ParsedPlayer;
  content: string[];
}

export interface ParseResult {
  logLines: ParsedLine[];
}

interface ParsedHeader {
  player: ParsedPlayer;
  time?: string;
}

type HeaderParser = (line: string) => ParsedHeader | null;
type LogLineConverter = (logLine: ParsedLine) => ParsedLine | null;

interface LogConfig {
  headerParser: HeaderParser;
  logLineConverter: LogLineConverter;
}

function removeSystemTextConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const withdrawParser = /^.*撤回了一条消息( 重新编辑)?$/;
  const withdrawMemberParser = /^.*撤回了成员.*的一条消息$/;
  const withdrawMember2Parser = /^.*撤回了一条成员消息$/;
  const commonFriendsParser = /^你和.*有\d+个共同好友，点击添加好友。$/;
  const joinMemberParser = /^.*加入本群。$/;
  const inviteMemberParser = /^.*邀请.*加入了本群。$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      withdrawParser,
      withdrawMemberParser,
      withdrawMember2Parser,
      commonFriendsParser,
      joinMemberParser,
      inviteMemberParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

const regNumber = /\(\d+\)|<.+@.+\..+>/;
const regTitle = /(?:【(.{1,6})】)?/;
const regTime = /\d{1,2}:\d{2}:\d{2}(?: (?:AM|PM))?/;

function removeMessageManagerSystemTextConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const dateParser = /^ \d{4}-\d{2}-\d{2}$/;
  const withdrawParser = new RegExp(`^${regTime.source}.*撤回了一条消息$`);
  const withdrawMemberParser = new RegExp(`^${regTime.source}.*撤回了成员.*的一条消息$`);
  const withdrawMember2Parser = new RegExp(`^${regTime.source}.*撤回了一条成员消息$`);
  // const commonFriendsParser = /^你和.*有\d+个共同好友，点击添加好友。$/;
  const joinMemberParser = new RegExp(`^${regTime.source}.*加入本群。$`);
  // const inviteMemberParser = /^.*邀请.*加入了本群。$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      dateParser,
      withdrawParser,
      withdrawMemberParser,
      withdrawMember2Parser,
      joinMemberParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

function defaultConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;
  // Remove starting and trailing empty lines.
  while(lines.length > 0 && lines[0].trim() === '') {
    lines.splice(0, 1);
  }
  while(lines.length > 0 && lines[lines.length-1].trim() === '') {
    lines.splice(lines.length-1, 1);
  }
  if (lines.length === 0)
    return null;
  return logLine;
}

function trimNumber(src: string): string {
  return trim(src, '()<>');
}

// Export from log
// E.g. "2019-09-23 8:43:38 PM 骰娘-Roll100(872001750)"
const exportFromLog: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regDateName = new RegExp(`\\d{4}-\\d{2}-\\d{2} ${regTime.source}`);
    const regHeader = new RegExp(`^(${regDateName.source}) (.*)(${regNumber.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, time, name, number] = matches;
    return {
      player: {
        name,
        number: trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    if (['10000', '1000000'].includes(logLine.player.number || ''))
      return null;
    return defaultConverter(logLine);
  },
};

// Copy from sidewindow
// E.g. "【冒泡】无情的围观熊 2/14/2020 9:16:13 PM"
const copyFromSideWindow: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regDateTime = new RegExp(`\\d{1,4}\\/\\d{1,2}\\/\\d{1,4} ${regTime.source}`);
    const regHeader = new RegExp(`^${regTitle.source}(.*?) (${regDateTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, time] = matches;
    return {
      player: {
        name,
        title,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from chat
// E.g. "【煤油】丧 丧 熊 9:59:54 PM"
const copyFromChat: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = new RegExp(`^${regTitle.source}(.*) (${regTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, time] = matches;
    return {
      player: {
        name,
        title,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from message manager
// E.g. "织练取(958884) 3:13:28 AM"
const copyFromMessageManager: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = new RegExp(`^${regTitle.source}(.*?)(${regNumber.source}) (${regTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, number, time] = matches;
    return {
      player: {
        name,
        title,
        number: trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeMessageManagerSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from mobile
// E.g. "a dark ideation  23:50:40"
const copyFromMobile: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = /^(.*?) {2}(\d{1,2}:\d{2}:\d{2}) *$/;
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, time] = matches;
    return {
      player: {
        name,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return defaultConverter(logLine);
  },
};

export function parseChat(data: string): ParseResult {
  const logLines: ParsedLine[] = [];
  let firstLogConfig: LogConfig | undefined = undefined;
  for (const line of data.split('\n')) {
    const parsedHeader = ((): ParsedHeader | null => {
      if (firstLogConfig)
        return firstLogConfig.headerParser(line);
      for (const logConfig of [
        copyFromMessageManager,
        copyFromMobile,
        exportFromLog,
        copyFromSideWindow,
        copyFromChat,
      ]) {
        const result = logConfig.headerParser(line);
        if (result) {
          firstLogConfig = logConfig;
          return result;
        }
      }
      return null;
    })();
    if (parsedHeader) {
      logLines.push({
        ...parsedHeader,
        content: [],
      });
    } else {
      if (firstLogConfig) {
        const currentLogLine: ParsedLine | undefined = logLines[logLines.length - 1];
        // Content after a recognizable header
        if (currentLogLine)
          currentLogLine.content.push(line);
      } else {
        // Unrecognizable content before the first header
        // Do nothing
      }
    }
  }
  if (!firstLogConfig)
    return { logLines: [] };

  const convertedLogLines = compact(
    logLines.map(firstLogConfig.logLineConverter)
  );
  return { logLines: convertedLogLines };
}