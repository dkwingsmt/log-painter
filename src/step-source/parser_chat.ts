import compact from 'lodash/compact';
import trim from 'lodash/trim';
import flow from 'lodash/flow';

export interface ParsedPlayer {
  name: string;
  number?: string;
  title?: string;
}

export interface ParsedLine {
  time?: string;
  player: ParsedPlayer;
  content: string[];
}

export interface ParseResult {
  logLines: ParsedLine[];
}

interface ParsedHeader {
  player: ParsedPlayer;
  time?: string;
  content?: string;
}

type HeaderParser = (line: string) => ParsedHeader | null;
type LogLineConverter = (logLine: ParsedLine) => ParsedLine | null;

interface LogConfig {
  headerParser: HeaderParser;
  logLineConverter: LogLineConverter;
}

function removeSystemTextConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const withdrawParser = /^.*撤回了一条消息/;
  const withdrawEnglishParser = /^.* recalled a message( Re-edit)?$/;
  const withdrawMemberParser = /^.*撤回了成员.*的一条消息$/;
  const withdrawMember2Parser = /^.*撤回了一条成员消息$/;
  const commonFriendsParser = /^你和.*有\d+个共同好友，点击添加好友。$/;
  const joinMemberParser = /^.*加入本群。$/;
  const inviteMemberParser = /^.*邀请.*加入了本群。$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      withdrawParser,
      withdrawEnglishParser,
      withdrawMemberParser,
      withdrawMember2Parser,
      commonFriendsParser,
      joinMemberParser,
      inviteMemberParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

const regNumber = /\(\d+\)|<.+@.+\..+>/;
const regTitle = /(?:【(.{1,6})】)?/;
const regTime = /(?:(?:上午|下午) )?\d{1,2}:\d{2}:\d{2}(?: (?:AM|PM))?/;

function removeMobileDateConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const dateParser = /^— —  \d{4}-\d{1,2}-\d{1,2}  — —$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      dateParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

function removeMessageManagerSystemTextConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const dateParser = /^ \d{4}-\d{2}-\d{2}$/;
  const withdrawParser = new RegExp(`^${regTime.source}.*撤回了一条消息`);
  const withdrawEnglishParser = new RegExp(`^${regTime.source}.* recalled a message( Re-edit)?`);
  const withdrawMemberParser = new RegExp(`^${regTime.source}.*撤回了成员.*的一条消息$`);
  const withdrawMember2Parser = new RegExp(`^${regTime.source}.*撤回了一条成员消息$`);
  // const commonFriendsParser = /^你和.*有\d+个共同好友，点击添加好友。$/;
  const joinMemberParser = new RegExp(`^${regTime.source}.*加入本群。$`);
  // const inviteMemberParser = /^.*邀请.*加入了本群。$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      dateParser,
      withdrawParser,
      withdrawEnglishParser,
      withdrawMemberParser,
      withdrawMember2Parser,
      joinMemberParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

function defaultConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;
  // Remove starting and trailing empty lines.
  while(lines.length > 0 && lines[0].trim() === '') {
    lines.splice(0, 1);
  }
  while(lines.length > 0 && lines[lines.length-1].trim() === '') {
    lines.splice(lines.length-1, 1);
  }
  if (lines.length === 0)
    return null;
  return logLine;
}

function trimNumber(src: string): string {
  return trim(src, '()<>');
}

// Export from log
// E.g. "2019-09-23 8:43:38 PM 骰娘-Roll100(872001750)"
// E.g. "2019-09-23 8:43:38 PM 骰娘-Roll100"
const exportFromLog: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regDateName = new RegExp(`\\d{4}-\\d{2}-\\d{2} ${regTime.source}`);
    const regHeader = new RegExp(`^(${regDateName.source}) (.*?)(${regNumber.source})?$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, time, name, number] = matches;
    return {
      player: {
        name,
        number: number == null ? undefined : trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    if (['10000', '1000000'].includes(logLine.player.number || ''))
      return null;
    return defaultConverter(logLine);
  },
};

// Copy from chat mainwindow or sidewindow
// E.g. "【冒泡】无情的围观熊 2/14/2020 9:16:13 PM"
// E.g. "【煤油】丧 丧 熊 9:59:54 PM"
const copyFromChat: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regDateTime = new RegExp(`(?:\\d{1,4}\\/\\d{1,2}\\/\\d{1,4} (?:星期. )?)?${regTime.source}`);
    const regHeader = new RegExp(`^${regTitle.source}(.*?) (${regDateTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, time] = matches;
    return {
      player: {
        name,
        title,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from message manager
// E.g. "织练取(958884) 3:13:28 AM"
const copyFromMessageManager: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = new RegExp(`^${regTitle.source}(.*?)(${regNumber.source}) (${regTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, number, time] = matches;
    return {
      player: {
        name,
        title,
        number: trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeMessageManagerSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from mobile
// E.g. "a dark ideation  23:50:40"
const copyFromMobile: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = /^(.*?) {2}(\d{1,2}:\d{2}:\d{2}) *$/;
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, time] = matches;
    return {
      player: {
        name,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeMobileDateConverter,
      defaultConverter,
    )(logLine);
  },
};

// 溯洄骰的Log记录器
// E.g. "名字(123456) 2021-11-08 04:07:37"
const shikiLogRecord: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = new RegExp(`^(.*?)(${regNumber.source}) (\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, number, time] = matches;
    return {
      player: {
        name,
        number: trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      defaultConverter,
    )(logLine);
  },
};

// Reparse log with angular brackes
// E.g. "<小明> 内容内容内容内容"
const reparseAngularBracket: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = /^<(.+)> (.+)$/;
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, content] = matches;
    return {
      player: {
        name,
      },
      content,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      defaultConverter,
    )(logLine);
  },
};

// Reparse log with bold brackes
// E.g. "【小明】内容内容内容内容"
const reparseBoldBracket: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = /^【(.+)】(.+)$/;
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, content] = matches;
    return {
      player: {
        name,
      },
      content,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      defaultConverter,
    )(logLine);
  },
};

export function parseChat(data: string): ParseResult {
  const logLines: ParsedLine[] = [];
  let firstLogConfig: LogConfig | undefined = undefined;
  for (const line of data.split('\n')) {
    const parsedHeader = ((): ParsedHeader | null => {
      if (firstLogConfig)
        return firstLogConfig.headerParser(line);
      for (const logConfig of [
        shikiLogRecord,
        copyFromMessageManager,
        copyFromMobile,
        exportFromLog,
        copyFromChat,
        reparseAngularBracket,
        reparseBoldBracket,
      ]) {
        const result = logConfig.headerParser(line);
        if (result) {
          firstLogConfig = logConfig;
          return result;
        }
      }
      return null;
    })();
    if (parsedHeader) {
      const { player, time, content } = parsedHeader;
      logLines.push({
        player,
        time,
        content: content ? [content] : [],
      });
    } else {
      if (firstLogConfig) {
        const currentLogLine: ParsedLine | undefined = logLines[logLines.length - 1];
        // Content after a recognizable header
        if (currentLogLine)
          currentLogLine.content.push(line);
      } else {
        // Unrecognizable content before the first header
        // Do nothing
      }
    }
  }
  if (!firstLogConfig)
    return { logLines: [] };

  const convertedLogLines = compact(
    logLines.map(firstLogConfig.logLineConverter)
  );
  return { logLines: convertedLogLines };
}
