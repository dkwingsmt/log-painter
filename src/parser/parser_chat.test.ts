import { parseChat } from './parser_chat';

it("header1-basic", () => {
  const text = `
消息记录（此消息记录为文本格式，不支持重新导入）

================================================================
消息分组:我的群聊
================================================================
消息对象:湖之仆从（PST 8pm | BJT 11am）
================================================================

2019-09-23 8:43:38 PM 骰娘-Roll100(87200175)
1

2019-09-23 8:43:38 PM 系统消息(10000)
[图片]2

2019-09-23 8:44:09 PM 系统消息(1000000)
3
4

2019-09-23 8:46:18 PM 暗夜(28084825)


2019-09-23 8:49:12 PM 系统消息(1000000)
真核生物独幕剧加入本群。

2019-09-23 9:02:42 PM 暗夜(28084825)
.r test

2019-09-23 11:26:00 PM 岁寒(36226407)
4

2019-09-24 7:54:29 AM 黄山真君(106505723)
5

2019-09-24 9:16:35 AM 黄山真君(106505723)
6
  `;
  expect(
    parseChat(text)
  ).toMatchSnapshot();
});

it("header2-basic", () => {
  const text = `
【冒泡】无情的围观熊 2/14/2020 9:16:13 PM
（1

【骰子】守善锤 2/14/2020 9:16:17 PM
2
3

你和abc有12个共同好友，点击添加好友。

【PL】克里斯托夫·韦恩 2/14/2020 9:17:10 PM
3

你撤回了一条消息

【PL】克里斯托夫·韦恩 2/14/2020 9:17:22 PM

守善锤撤回了一条消息

【PL】菲比斯·墨菲斯托 2/14/2020 9:17:31 PM
5
  `;
  expect(
    parseChat(text)
  ).toMatchSnapshot();
});

it("header2-without title", () => {
  const text = `
无情的围观熊 2/14/2020 9:16:13 PM
（1

守善锤 2/14/2020 9:16:17 PM
2
3

你和abc有12个共同好友，点击添加好友。

克里斯托夫·韦恩 2/14/2020 9:17:10 PM
3

你撤回了一条消息

克里斯托夫·韦恩 2/14/2020 9:17:22 PM

守善锤撤回了一条消息

菲比斯·墨菲斯托 2/14/2020 9:17:31 PM
5
  `;
  expect(
    parseChat(text)
  ).toMatchSnapshot();
});

it("header3-basic", () => {
  const text = `
【煤油】丧 丧 熊 9:59:54 PM
1

【煤油】丧 丧 熊 9:59:54 PM

守善锤撤回了一条消息

你和abc有12个共同好友，点击添加好友。

【八方体】Paul~K（写不完了...） 10:00:04 PM
2
3
  `;
  expect(
    parseChat(text)
  ).toMatchSnapshot();
});

it("header3-without title", () => {
  const text = `
丧 丧 熊 9:59:54 PM
1

丧 丧 熊 9:59:54 PM

守善锤撤回了一条消息

你和abc有12个共同好友，点击添加好友。

Paul~K（写不完了...） 10:00:04 PM
2
3
  `;
  expect(
    parseChat(text)
  ).toMatchSnapshot();
});

it("header3-instant withdraw", () => {
  const text = `
丧 丧 熊 9:59:54 PM
1

你撤回了一条消息 重新编辑

Paul~K（写不完了...） 10:00:04 PM
2
3
  `;
  expect(
    parseChat(text)
  ).toMatchSnapshot();
});

it("header4-basic", () => {
  const text = `
— —  2020-2-21  — —

德里奇化石  23:50:12${" "}
1

a dark ideation  23:50:40${" "}
2

一零绵羊咩咩咩  23:50:48${" "}
诶？
  `;
  expect(
    parseChat(text)
  ).toMatchSnapshot();
});