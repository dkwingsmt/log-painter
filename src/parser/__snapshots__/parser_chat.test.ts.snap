// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copied from chat basic: datetime format 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "（1",
      ],
      "player": Object {
        "name": "无情的围观熊",
        "title": "冒泡",
      },
      "time": "2/14/2020 9:16:13 PM",
    },
    Object {
      "content": Array [
        "2",
        "3",
      ],
      "player": Object {
        "name": "守善锤",
        "title": "骰子",
      },
      "time": "2/14/2020 9:16:17 PM",
    },
    Object {
      "content": Array [
        "3",
      ],
      "player": Object {
        "name": "克里斯托夫·韦恩",
        "title": "PL",
      },
      "time": "2/14/2020 9:17:10 PM",
    },
    Object {
      "content": Array [
        "5",
      ],
      "player": Object {
        "name": "菲比斯·墨菲斯托",
        "title": "PL",
      },
      "time": "2/14/2020 9:17:31 PM",
    },
  ],
}
`;

exports[`copied from chat basic: time format 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "丧 丧 熊",
        "title": "煤油",
      },
      "time": "9:59:54 PM",
    },
    Object {
      "content": Array [
        "2",
        "3",
      ],
      "player": Object {
        "name": "Paul~K（写不完了...）",
        "title": "八方体",
      },
      "time": "10:00:04 PM",
    },
  ],
}
`;

exports[`copied from chat date format without AM|PM 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "伤情",
      ],
      "player": Object {
        "name": "迈克尔 史密斯",
        "title": "酒店管理层",
      },
      "time": "2020/3/28 19:02:31",
    },
  ],
}
`;

exports[`copied from chat instant withdraw 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "丧 丧 熊",
        "title": undefined,
      },
      "time": "9:59:54 PM",
    },
    Object {
      "content": Array [
        "2",
        "3",
      ],
      "player": Object {
        "name": "Paul~K（写不完了...）",
        "title": undefined,
      },
      "time": "10:00:04 PM",
    },
  ],
}
`;

exports[`copied from chat mixed date-time format 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "爬行类化石",
        "title": undefined,
      },
      "time": "5/8/2020 11:59:47 PM",
    },
    Object {
      "content": Array [
        "2",
      ],
      "player": Object {
        "name": " Patrick",
        "title": undefined,
      },
      "time": "12:03:01 AM",
    },
    Object {
      "content": Array [
        "3",
      ],
      "player": Object {
        "name": "爬行类化石",
        "title": undefined,
      },
      "time": "12:07:49 AM",
    },
  ],
}
`;

exports[`copied from chat other system messages 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "守善锤",
        "title": "骰子",
      },
      "time": "3/4/2020 6:00:32 PM",
    },
    Object {
      "content": Array [
        "2",
      ],
      "player": Object {
        "name": "克里斯托夫·韦恩",
        "title": "嘟嘟男爵",
      },
      "time": "3/4/2020 6:00:37 PM",
    },
    Object {
      "content": Array [
        "3",
      ],
      "player": Object {
        "name": "薇薇安·汉密尔顿",
        "title": "PL",
      },
      "time": "3/4/2020 6:01:02 PM",
    },
  ],
}
`;

exports[`copied from chat single date 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "守善锤",
        "title": "骰子",
      },
      "time": "3/4/2020 6:00:32 PM",
    },
    Object {
      "content": Array [
        "2",
      ],
      "player": Object {
        "name": "克里斯托夫·韦恩",
        "title": "嘟嘟男爵",
      },
      "time": "3/4/2020 6:00:37 PM",
    },
    Object {
      "content": Array [
        "3",
      ],
      "player": Object {
        "name": "薇薇安·汉密尔顿",
        "title": "PL",
      },
      "time": "3/4/2020 6:01:02 PM",
    },
  ],
}
`;

exports[`copied from chat without title 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "丧 丧 熊",
        "title": undefined,
      },
      "time": "9:59:54 PM",
    },
    Object {
      "content": Array [
        "2",
        "3",
      ],
      "player": Object {
        "name": "Paul~K（写不完了...）",
        "title": undefined,
      },
      "time": "10:00:04 PM",
    },
  ],
}
`;

exports[`copied from message manager Chinese time 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "“——1。",
      ],
      "player": Object {
        "name": "织练取",
        "number": "958884",
        "title": undefined,
      },
      "time": "上午 3:13:28",
    },
    Object {
      "content": Array [
        "“我2”",
        "\\"门3\\"",
      ],
      "player": Object {
        "name": "白菜",
        "number": "indo@qq.com",
        "title": undefined,
      },
      "time": "上午 3:14:30",
    },
    Object {
      "content": Array [
        "嫌疑",
        "是出",
      ],
      "player": Object {
        "name": "白菜",
        "number": "indo@qq.com",
        "title": undefined,
      },
      "time": "下午 3:16:49",
    },
  ],
}
`;

exports[`copied from message manager basic 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "“——1。",
      ],
      "player": Object {
        "name": "织练取",
        "number": "958884",
        "title": "1",
      },
      "time": "3:13:28 AM",
    },
    Object {
      "content": Array [
        "“我2”",
        "\\"门3\\"",
      ],
      "player": Object {
        "name": "白菜",
        "number": "indo@qq.com",
        "title": "2",
      },
      "time": "3:14:30 AM",
    },
    Object {
      "content": Array [
        "嫌疑",
        "是出",
      ],
      "player": Object {
        "name": "白菜",
        "number": "indo@qq.com",
        "title": "2",
      },
      "time": "3:16:49 AM",
    },
  ],
}
`;

exports[`copied from message manager system messages 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "暗夜桑的h检定，那么结果是:D100=13/25 成功",
      ],
      "player": Object {
        "name": "骰娘",
        "number": "87200",
        "title": undefined,
      },
      "time": "4:11:31 PM",
    },
    Object {
      "content": Array [
        ".r4h",
      ],
      "player": Object {
        "name": "DK",
        "number": "2808",
        "title": undefined,
      },
      "time": "12:11:41 AM",
    },
  ],
}
`;

exports[`copied from message manager without title 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "“——1。",
      ],
      "player": Object {
        "name": "织练取",
        "number": "958884",
        "title": undefined,
      },
      "time": "3:13:28 AM",
    },
    Object {
      "content": Array [
        "“我2”",
        "\\"门3\\"",
      ],
      "player": Object {
        "name": "白菜",
        "number": "indo@qq.com",
        "title": undefined,
      },
      "time": "3:14:30 AM",
    },
    Object {
      "content": Array [
        "嫌疑",
        "是出",
      ],
      "player": Object {
        "name": "白菜",
        "number": "indo@qq.com",
        "title": undefined,
      },
      "time": "3:16:49 AM",
    },
  ],
}
`;

exports[`copied from mobile basic 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "德里奇化石",
      },
      "time": "23:50:12",
    },
    Object {
      "content": Array [
        "2",
      ],
      "player": Object {
        "name": "a dark ideation",
      },
      "time": "23:50:40",
    },
    Object {
      "content": Array [
        "诶？",
      ],
      "player": Object {
        "name": "一零绵羊咩咩咩",
      },
      "time": "23:50:48",
    },
  ],
}
`;

exports[`copied from mobile filter dates 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "德里奇化石",
      },
      "time": "23:50:12",
    },
    Object {
      "content": Array [
        "2",
      ],
      "player": Object {
        "name": "a dark ideation",
      },
      "time": "23:50:40",
    },
  ],
}
`;

exports[`exported from log basic 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "骰娘-Roll100",
        "number": "87200175",
      },
      "time": "2019-09-23 8:43:38 PM",
    },
    Object {
      "content": Array [
        ".r test",
      ],
      "player": Object {
        "name": "暗夜",
        "number": "28084825",
      },
      "time": "2019-09-23 9:02:42 PM",
    },
    Object {
      "content": Array [
        "4",
      ],
      "player": Object {
        "name": "岁寒",
        "number": "36226407",
      },
      "time": "2019-09-23 11:26:00 PM",
    },
    Object {
      "content": Array [
        "5",
      ],
      "player": Object {
        "name": "黄山真君",
        "number": "106505723",
      },
      "time": "2019-09-24 7:54:29 AM",
    },
    Object {
      "content": Array [
        "6",
      ],
      "player": Object {
        "name": "黄山真君",
        "number": "106505723",
      },
      "time": "2019-09-24 9:16:35 AM",
    },
  ],
}
`;

exports[`exported from log email 1`] = `
Object {
  "logLines": Array [
    Object {
      "content": Array [
        "1",
      ],
      "player": Object {
        "name": "礼",
        "number": "852649",
      },
      "time": "2020-03-04 12:12:29 AM",
    },
    Object {
      "content": Array [
        "2",
      ],
      "player": Object {
        "name": "kp",
        "number": "100550",
      },
      "time": "2020-03-04 12:12:39 AM",
    },
    Object {
      "content": Array [
        "3",
      ],
      "player": Object {
        "name": "围观",
        "number": "indo@qq.com",
      },
      "time": "2020-03-04 12:13:41 AM",
    },
  ],
}
`;
